Adding boost as a submodule rather than depending on external references.

clemahieu opened the pull request:
Updating to boost 1.81 and removing CMake target links to Boost::boost which are no longer needed

Added boost_checkout_lite.sh which can be used to only clone required dependencies rather than every boost submodule in the /libs folder. The boost submodule can be initialized with: git submodule update --init boost && ./boost_checkout_lite.sh

clemahieu commented:
Related to https://github.com/nanocurrency/nano-node/pull/4024

clemahieu commented:
Disk usage stats:

nano-node repo without boost submodule initialised
nano-node.src % du -sh .
1.5G	.

Size of full boost repo:
boost.src % du -sh .
1.9G

Boost source extracted from archive:
boost_1_74_0$ du -sh .
798M

With boost_checkout_lite.sh
nano-node.src % du -sh .
2.6G

With full recursive checkout
nano-node.src % du -sh .
3.5G

pwojcikdev commented on nano/lib/logger_mt.hpp:
```diff
@@ -134,6 +134,6 @@ class logger_mt
 	nano::mutex last_log_time_mutex;
 	std::chrono::steady_clock::time_point last_log_time;
-	boost::log::sources::severity_logger_mt<severity_level> boost_logger_mt;
+	boost::log::sources::severity_logger<severity_level> boost_logger_mt;
```
This version of logger (without _mt) is not thread safe and we're using it from multiple threads.

dsiganos commented on CMakeLists.txt:
```diff
@@ -591,7 +663,7 @@ if(NANO_FUZZER_TEST)
 
 if(NANO_TEST OR RAIBLOCKS_TEST)
-  find_package(Boost 1.70.0 REQUIRED COMPONENTS coroutine context)
+  # find_package(Boost 1.70.0 REQUIRED COMPONENTS coroutine context)
```
This can be removed

dsiganos commented on CMakeLists.txt:
```diff
@@ -363,35 +363,106 @@ else()
 
-find_package(Boost 1.70.0 REQUIRED COMPONENTS filesystem log log_setup thread
-                                              program_options system)
+# find_package(Boost 1.70.0 REQUIRED COMPONENTS filesystem log log_setup thread
```
the version v1.70 n the comment makes no sense anymore

dsiganos commented on CMakeLists.txt:
```diff
@@ -485,6 +556,7 @@ add_definitions(-DCRYPTOPP_DISABLE_SHANI)
 # Fix failing builds after commit
 # https://github.com/weidai11/cryptopp/commit/df9fa62205f2d341e2b1b26595a3a1b6377c60c5
 add_definitions(-DCRYPTOPP_DISABLE_CLMUL)
+add_definitions(-DCRYPTOPP_DISABLE_ASM)
```
If we are going disable functionality that otherwise looks good to have then we need explain why we are disabling it.

clemahieu replied:
This was a workaround for cross-compiling, I'll remove it.

dsiganos commented on ci/actions/windows/build.ps1:
```diff
@@ -41,7 +39,7 @@ mkdir build
 
 #accessibility of Boost dlls for generating config samples
-$ENV:PATH = "$ENV:PATH;$ENV:BOOST_ROOT\lib64-msvc-14.2"
+$ENV:PATH = "$ENV:PATH"
```
Is this line needed at all?

dsiganos commented on nano/node/logging.cpp:
```diff
@@ -37,20 +36,6 @@ void nano::logging::init (boost::filesystem::path const & application_path_a)
 		}
 
 #ifdef BOOST_WINDOWS
-		if (nano::event_log_reg_entry_exists () || nano::is_windows_elevated ())
```
Does the removal of this code mean that on windows we do not log to file?

clemahieu replied:
This means we can't log to the Windows Event Log program, a kind of Windows syslog. I'll raise an issue to reimplement this if we can get the GitHub builders to support the required tool, the Windows Message Compiler mc.ex.. Short of this I don't know anyone that currently uses Windows + Event Log so I wouldn't hold up the boost submodule conversion on this point.

dsiganos replied:
Windows is our least supported platform, I do not think we should have extra features on the platform we support the least.
As long as logging to file exists on windows, I don't think we should do anymore than that.

clemahieu committed:
Update boost requirement to 1.81.0 and fix some API changes in rpc_connection.cpp

clemahieu committed:
Adding boost as a submodule rather than requiring it to be provided externally.
The boost libs are now statically linked to the nano targets so dynamic linking related functionality has been removed.
Adding several target link references that were missing but not apparent when using the unified boost install directories
Linking to boost stacktrace basic as this is the main functionality we use.
Windows Event Log support has been disabled as there is a compatibility issue with GitHub builders
Windows cryptopp ASM has been disabled since CMake isn't assembling .asm files for unknown reasons.
Added boost_checkout_lite.sh script which can be used to do a minimal clone of the boost submodule after it's been added but not yet initialized.

pwojcikdev reviewed (dismissed):
It's good to see the conversion removed a lot of unnecessary complexity from cmakefiles and source
